synapse/storage/databases/main/event_push_actions.py

```diff
@@ -1083,7 +1199,7 @@ def _handle_new_receipts_for_notifs_txn(self, txn: LoggingTransaction) -> bool:
         )
 
         sql = """
-            SELECT r.stream_id, r.room_id, r.user_id, e.stream_ordering
+            SELECT r.stream_id, r.room_id, r.user_id, r.thread_id, e.stream_ordering
```

Comment:
Why do we pull out `r.stream_id` if we never use it in the loops below?

Reply:
Ah, I noticed this and meant to do it as a separate PR (unrelated to this work). I'll split that out now.

Reply:
See #13933 -- I'm going to mark this as resolved and handle it in that PR.

Reply:
Is there a way we can make this more obvious since we all missed it? 😅 Harder to see in the diff I guess.

The best I am coming up with is this sort of stuff:

```suggestion
            SELECT
                /**
                 * We use this after we're done processing to mark the last receipt that we processed
                 * (see `receipts_last_processed_stream_id`)
                 */
                r.stream_id,
                /* We pull the rest of these out to use while processing */
                r.room_id, r.user_id, r.thread_id, e.stream_ordering
            ...
```


Renaming `upper_limit` to `receipts_last_processed_stream_id` seems like a separate thing to make the code more clear.

Reply:
I was thinking of this and couldn't really come up with much, unfortunately.

Maybe putting the check *before* the large for-loop would make it clearer? Or adding proper type hints would at least have made it error?

Reply:
> Maybe putting the check before the large for-loop would make it clearer?

It would be more clear but it does make sense next to the usage as well. I don't have a strong leaning to actually do this though.

> Or adding proper type hints would at least have made it error?

Adding type hints here would be great. I'm curious how that would look

Reply:
https://github.com/matrix-org/synapse/compare/clokep/receipt-types?expand=0 has the type hints, it is just a `cast()` on what `rows` is.

I did check though that if you remove the `stream_id` (and the corresponding first `int` in the cast) than mypy would fail. So that does seem useful?

Reply:
If it catches some things, seems good 👍. Some brittleness having to keep both in sync.

Only other thought is a `_ReceiptInfo` type so we can use a named lookup, `receipts_last_processed_stream_id = receipt_infos[-1].stream_id`

Reply:
This was done in #13939 and this PR now shows the change to the return value.